src/widget/wfindonwebmenu.cpp

```diff
@@ -0,0 +1,163 @@
+
+const QString kSearchUrlLastFmAlbum = QStringLiteral("https://www.last.fm/search/albums?");
+
+const QString kSearchUrlDefault = QStringLiteral("https://soundcloud.com/search?");
+
+QString getServiceUrl(WFindOnWebMenu::Service service,
+        WFindOnWebMenu::TrackSearchProperties trackSearchProperty) {
+    switch (service) {
+    case WFindOnWebMenu::Service::Discogs:
+        return kSearchUrlDiscogsGen;
+    case WFindOnWebMenu::Service::LastFm:
+        if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Title ||
+                trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle) {
+            return kSearchUrlLastFmTitle;
+        } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Artist) {
+            return kSearchUrlLastFmArtist;
+        } else {
+            return kSearchUrlLastFmAlbum;
+        }
+    case WFindOnWebMenu::Service::SoundCloud:
+        if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Title ||
+                trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle) {
+            return kSearchUrlSoundCloudTitle;
+        } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Artist) {
+            return kSearchUrlSoundCloudArtist;
+        } else {
+            return kSearchUrlSoundCloudAlbum;
+        }
+    default:
+        return kSearchUrlDefault;
+    }
+}
+} // namespace
+
+WFindOnWebMenu::WFindOnWebMenu(QWidget* parent)
+        : QMenu(tr("Find on Web"), parent) {
+}
+
+void WFindOnWebMenu::addSubmenusForServices(const Track& track) {
+    WFindOnWebMenu::populateFromTrackProperties(
+            track,
+            QString(tr("SoundCloud")),
+            Service::SoundCloud);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("LastFm")), Service::LastFm);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("Discogs")), Service::Discogs);
+}
+
+void WFindOnWebMenu::addActions(Service service,
+        const QString& trackProperty,
+        QMenu* pServiceMenu,
+        TrackSearchProperties trackSearchProperty) {
+    const QString prefixAction = composePrefixAction(trackSearchProperty);
+    pServiceMenu->addAction(
+            composeActionText(prefixAction, trackProperty),
+            this,
+            [this, service, trackSearchProperty, trackProperty]() {
+                emit triggerBrowser(service, trackSearchProperty, trackProperty);
+            });
+}
+
+void WFindOnWebMenu::openInBrowser(Service service,
+        TrackSearchProperties trackSearchProperty,
+        const QString& query) {
+    const QString serviceUrl = getServiceUrl(service, trackSearchProperty);
+    QUrlQuery urlQuery;
+    urlQuery.addQueryItem("q", query);
+    QUrl url(serviceUrl);
+    url.setQuery(urlQuery);
+    VERIFY_OR_DEBUG_ASSERT(url.isValid()) {
```

Comment:
The warning message is wrong at this point. QDesktopServices::openUrl() wasn't event invoked.

Reply:
openUrl() has a URL validation inside. [openUrl() URL validation](https://github.com/qt/qtbase/blob/42e13b7c61693ca95e119106d4f6dbd2bcf2308d/src/gui/util/qdesktopservices.cpp#L210)

So my check was duplication of an existing check and redundant. So instead, it is better to check if "QDesktopServices::openUrl()" fails itself and warn the developers with "DEBUG_ASSERT(false)", so it logged before Mixxx quits the assertion.

Also about the first check, I was checking if the URL is empty, this wasn't the best validation for Url's and also it was redundant too.  

I will now commit the URL validation again, I hope this time we will have a well working URL validation.